pyomo/contrib/gdpopt/solve_master_problem.py

```diff
@@ -0,0 +1,185 @@
+            config.logger.warning(
+                "GAMS encountered an error in solve. Treating as infeasible.")
+            return tc.infeasible
+        else:
+            raise
+
+    config.call_after_master_solve(solver, m, util_block)
+
+    terminate_cond = results.solver.termination_condition
+    if terminate_cond is tc.infeasibleOrUnbounded:
+        # Linear solvers will sometimes tell me that it's infeasible or
+        # unbounded during presolve, but fails to distinguish. We need to
+        # resolve with a solver option flag on.
+        results, terminate_cond = distinguish_mip_infeasible_or_unbounded(
+            m, config)
+    if terminate_cond is tc.unbounded:
+        # Solution is unbounded. This occurs when the objective is
+        # nonlinear. The nonlinear objective is moved to the constraints, and
+        # deactivated for the linear master problem. We will generate an
+        # arbitrary discrete solution by bounding the objective and re-solving,
+        # in hopes that the cuts we generate later bound this problem.
+
+        obj_bound = 1E15
+        config.logger.warning(
+            'Master problem was unbounded. '
+            'Re-solving with arbitrary bound values of (-{0:.10g}, {0:.10g}) '
+            'on the objective, in order to get a discrete solution. '
+            'Check your initialization routine.'.format(obj_bound))
+        main_objective = next(m.component_data_objects(Objective, active=True))
+        util_block.objective_bound = Constraint(
+            expr=(-obj_bound, main_objective.expr, obj_bound))
+        with SuppressInfeasibleWarning():
+            results = SolverFactory(config.mip_solver).solve(
+                m, **config.mip_solver_args)
+        # get rid of the made-up constraint
+        del util_block.objective_bound
+        if results.solver.termination_condition in {tc.optimal, tc.feasible,
+                                                    tc.locallyOptimal,
+                                                    tc.globallyOptimal}:
+            # we found a solution, that's all we need to keep going.
+            return tc.unbounded
+        else:
+            raise RuntimeError("Unable to find a feasible solution for the "
+                               "unbounded MILP master problem by bounding "
+                               "the objective. Either check your "
+                               "master problem initialization, or add a bound "
+                               "on the master objective value that admits a "
+                               "feasible solution.")
+
+    if terminate_cond is tc.optimal:
+        return tc.optimal
+    elif terminate_cond in {tc.locallyOptimal, tc.feasible}:
+        return tc.feasible
+    elif terminate_cond is tc.infeasible:
+        config.logger.info(
+            'MILP master problem is now infeasible. '
+            'GDPopt has finished exploring feasible discrete configurations.')
```

Comment:
*might* have finished exploring feasible discrete configurations. Invalid cuts coming from LOA might lead to infeasible master without enumerating.